File: CEOI/2001/ceoionline2001/Day1/chain/C/meter.c
#include "meter.h"
#include <stdio.h>
#include <stdlib.h>

#include <string.h>

#define MaxN            10000
#define Pr              999999001L
#define true 1
#define false 0

typedef int Map[MaxN];
static int Init = true;
static Test = false;
static int N;
static Map P, CP, Acc;
static FILE *MetinF, *OutF;
static int MaxA;
static long Sign;

static void Abort(char *M)
{
  fprintf(OutF, "%s\n", M);
  if (OutF != NULL)
    fclose(OutF);
  OutF = NULL;
  exit(0);
}

static void T1(int *P, int *Q)
{
  int x, xx, y, L, FORLIM;

  L = P[0] + 1;
  if (L > N)
    L = N;
  FORLIM = N;
  for (x = 1; x <= FORLIM; x++) {
    xx = P[x - 1];
    if (xx % L == 0) {
      y = xx - L + 1;
    } else if (xx < N)
      y = xx + 1;
    else
      y = xx / L * L + 1;
    Q[y - 1] = x;
  }
}

static void Initialise(void)
{
  long x, y, Sign0;
  Map Q;
  long FORLIM;
  int SETIO;

  OutF = fopen("chain.out", "w");  
  MetinF = NULL;
  MetinF = fopen("chain.in", "r");
  if (MetinF == NULL) 
    Abort("ERROR: Input file CHAIN.IN not found!");
  SETIO = fscanf(MetinF, "%hd", &N);

  if (SETIO == 0) {
    if (MetinF != NULL)
      fclose(MetinF);
    MetinF = NULL;
    Abort("ERROR: Non-number value in first line of CHAIN.IN");
  }
  fscanf(MetinF, "%*[^\n]");
  if (N < 3 || N > MaxN) {
    if (MetinF != NULL)
      fclose(MetinF);
    MetinF = NULL;
    Abort("ERROR: Value out of range in first line of CHAIN.IN");
  }

  FORLIM = N;
  for (x = 0; x < FORLIM; x++) {
    Q[x] = -1;
    CP[x] = 0;
    Acc[x] = 0;
  }
  FORLIM = N;
  for (x = 1; x <= FORLIM; x++) {
    SETIO = fscanf(MetinF, "%ld", &y);
    if (SETIO = 0) {
      if (MetinF != NULL)
	fclose(MetinF);
      MetinF = NULL;
      Abort("ERROR: Non-number value in second line of file CHAIN.IN");
    }
    if (y < 1 || y > N) {
      if (MetinF != NULL)
	fclose(MetinF);
      MetinF = NULL;
      Abort("ERROR: Value out of range in second line of file CHAIN.IN");
    }
    if (Q[y - 1] != -1) {
      if (MetinF != NULL)
	fclose(MetinF);
      MetinF = NULL;
      Abort("ERROR: Duplicate values in second line of file CHAIN.IN");
    }

    Q[y - 1] = x;
  }

  fscanf(MetinF, "%*[^\n]");
  Sign0 = -1;
  SETIO = fscanf(MetinF, "%ld", &Sign0);

  if (MetinF != NULL)
    fclose(MetinF);
  MetinF = NULL;

  Test = false;

  T1(Q, P);

  Sign = 0;
  FORLIM = N;
  for (x = 1; x <= FORLIM; x++) {
    y = P[x - 1];
    Sign = (Sign + (MaxN - x) * y) % Pr;
  }

  if (Sign != Sign0)
    memcpy(P, Q, sizeof(Map));
  else {
    Test = true;
    FORLIM = N;
    for (x = 1; x <= FORLIM; x++)
      Q[P[x - 1] - 1] = x;
    memcpy(P, Q, sizeof(Map));
  }

  MaxA = 0;
  Init = false;
}


int Size(void)
{
  int Result;

  if (Init)
    Initialise();

  fprintf(OutF, "Size=%d\n", N);
  fflush(OutF);
  return N;
}

void Answer(int i, int x)
{
  char Si[256], Sx[256];
  int ii;
  char STR2[256];
  int FORLIM;

  if (i < 1 || i > N || x < 1 || x > N) {
    sprintf(Si, "%d", i);
    sprintf(Sx, "%d", x);
    sprintf(STR2, "ERROR: in Answer(%s,%s)", Si, Sx);
    Abort(STR2);
  }
  CP[i - 1] = x;
  if (i != N)
    return;
  fprintf(OutF, "Your Answer:\n");
  FORLIM = N - 2;
  for (ii = 0; ii <= FORLIM; ii++)
    fprintf(OutF, "%d ", CP[ii]);
  fprintf(OutF, "%d\n", CP[N - 1]);
  fprintf(OutF, "Access/Atom:\n");
  fprintf(OutF, "%d\n", MaxA);
  if (Test) {
    fprintf(OutF, "Signature:\n");
    fprintf(OutF, "%ld\n", Sign);
  }
  if (OutF != NULL)
    fclose(OutF);
  OutF = NULL;
  exit(0);
}

int Span(int x, int y)
{
  int d;

  if (Init)
    Abort("ERROR: Illegal Dialogue! Size must be called first!");
  if (x < 1 || x > N || y < 1 || y > N) {
    fprintf(OutF, "Span(%d,%d)\n", x, y);
    fflush(OutF);
    Abort("ERROR: Parameter out of range in call Span!");
  }
  if (x == y)
    Acc[x - 1]++;
  else {
    Acc[x - 1]++;
    Acc[y - 1]++;
  }

  d = abs(P[x - 1] - P[y - 1]);
  fprintf(OutF, "Span(%d,%d)=%d\n", x, y, d);
  fflush(OutF);

  if (Acc[x - 1] > MaxA)
    MaxA = Acc[x - 1];
  if (Acc[y - 1] > MaxA)
    MaxA = Acc[y - 1];

  if (MaxA > 4) {
    OutF = fopen("chain.out", "w");
    fprintf(OutF, "%d\n\n", MaxA);
    if (Test)
      fprintf(OutF, "%ld\n", Sign);
    if (OutF != NULL)
      fclose(OutF);
    OutF = NULL;

    Abort("ERROR: Illegal Dialogue! Atom accessed more than four times.");
  }
  return d;
}

File: CEOI/2001/ceoionline2001/Day1/chain/P/meter.pas
Unit Meter;
Interface
  Function Size:Integer;
  Function Span(x,y:Integer):Integer;
  Procedure Answer(i:Integer; x:Integer);

Implementation
Const
  MaxN=10000;
  Init:Boolean=True;
  Test:Boolean=False;
  Pr=999999001;
Type
  Map=Array[1..MaxN] Of Integer;
Var
  N:Integer;
  P,
  CP,
  Acc:Map;
  MetinF,
  OutF:Text;
  MaxA:Integer;
  Sign:Longint;

Procedure Abort(M:String);
Begin
  WriteLn(OutF,M);
  Close(OutF);
  Halt;
End;

Procedure T1( P:Map; Var Q:Map);
  Var x,xx,y:Integer;
    L:Integer;
  Begin
    L:=P[1]+1; If L>N Then L:=N;
    For x:=1 To N Do Begin
      xx:=P[x];
      If (xx Mod L=0) Then
        y:=xx-L+1
      Else If xx<N Then
        y:=xx+1
      Else
        y:=(xx div L)*L+1;
      Q[y]:=x;
    End;
  End;

Procedure Initialise;
Var
  x,y:Longint;
  Sign0:Longint;
  Q:Map;
Begin
  Assign(OutF, 'chain.out');
  Rewrite(OutF);
  Assign(MetInF, 'chain.in');
  {$I-}
    Reset(MetInF);
  {$I+}
  If IOResult<>0 Then
    Abort('ERROR: Input file CHAIN.IN not found!');
  {$I-}
    ReadLn(MetInF,N);
  {$I+}
  If (IOResult<>0) Then  Begin
    Close(MetInF);
    Abort('ERROR: Non-number value in first line of CHAIN.IN');
  End;
  If (N<3) Or (N>MaxN) Then  Begin
    Close(MetInF);
    Abort('ERROR: Value out of range in first line of CHAIN.IN');
  End;

  For x:=1 To N Do Begin
    Q[x]:=-1;
    CP[x]:=0;
    ACC[x]:=0;
  End;
  For x:=1 To N Do Begin
    If SeekEoLn(MetInF) Then  Begin
      Close(MetInF);
      Abort('ERROR: Number of items in second line of file CHAIN.IN less then required');
    End;
    {$I-}
      Read(MetInF,y);
    {$I+}
    If (IOResult<>0) Then  Begin
      Close(MetInF);
      Abort('ERROR: Non-number value in second line of file CHAIN.IN');
    End;
    If (y<1) Or (y>N) Then  Begin
      Close(MetInF);
      Abort('ERROR: Value out of range in second line of file CHAIN.IN');
    End;
    If (Q[y]<>-1) Then  Begin
      Close(MetInF);
      Abort('ERROR: Duplicate values in second line of file CHAIN.IN');
    End;

    Q[y]:=x;
  End;

  ReadLn(MetInF);
  If EoF(MetInF) Then
    Sign0:=-1
  Else Begin
    {$I-}
      ReadLn(MetInF,Sign0);
    {$I+}
    If IOResult<>0 Then
      Sign0:=-1
  End;
  Close(MetInF);

  Test:=False;

  T1(Q,P);

  Sign:=0;
  For x:=1 To N Do Begin
    y:=P[x];
    Sign:=(Sign+(MaxN-x)*y) Mod Pr;
  End;

  If Sign<>Sign0 Then Begin
    P:=Q;
  End Else Begin
    Test:=True;
    For x:=1 To N Do
      Q[P[x]]:=x;
    P:=Q;
  End;

  MaxA:=0;
  Init:=False;
End{Initialise};

Function Size:Integer;
Begin
  If Init Then
    Initialise;
  Size:=N;
  WriteLn(OutF,'Size=',N:1);
  Flush(OutF);
End;

Procedure Answer(i:Integer;x:Integer);
Var
  Si,Sx:String;
  ii:Integer;
Begin
  If (i<1)Or(i>N)Or(x<1)Or(x>N) Then Begin
    Str(i,Si); Str(x,Sx);
    Abort('ERROR: in Answer('+Si+','+Sx+')');
  End;
  CP[i]:=x;
  If i=N Then Begin
    WriteLn(OutF,'Your Answer:');
    For ii:=1 To N-1 Do
      Write(OutF,CP[ii]:1,' ');
    WriteLn(OutF,CP[N]:1);
    WriteLn(OutF,'Access/Atom:');
    WriteLn(OutF,MaXA);
    If Test Then Begin
      WriteLn(OutF,'Signature:');
      WriteLn(OutF,Sign);
    End;
    Close(OutF);
    Halt;
  End;
End{Answer};

Function Span(x,y:Integer):Integer;
Var d:Integer;
Begin
  If Init Then
    Abort('ERROR: Illegal Dialogue! Size must be called first!');
  If (x<1) Or (x>N) Or (y<1) Or (y>N) Then Begin
    WriteLn(OutF,'Span(',x:1,',',y:1,')');
    Flush(OutF);
    Abort('ERROR: Parameter out of range in call Span!');
  End;
  If x=y Then
    Inc(Acc[x])
  Else Begin
    Inc(Acc[x]); Inc(Acc[y]);
  End;

  d:=Abs(P[x]-P[y]);
  WriteLn(OutF,'Span(',x:1,',',y:1,')=',d:1);
  Flush(OutF);
  If  (Acc[x]>MaxA) Then
    MaxA:=Acc[x];
  If  (Acc[y]>MaxA) Then
    MaxA:=Acc[y];

  If MaxA>4 Then Begin
    Assign(OutF, 'chain.out'); Rewrite(OutF);
    WriteLn(OutF,MaxA);
    WriteLn(OutF);

    Abort('ERROR: Illegal Dialogue! Atom accessed more than four times.');
  End;
  Span:=d;
End;

End.
